robotoff/insights/importer.py

```diff
@@ -283,354 +410,294 @@ def ignore_insight(
+    ) -> Iterator[ProductInsight]:
+        yield from (
+            ProductInsight(**prediction.to_dict())
+            for prediction in predictions
+            if not cls.ignore_prediction(product, prediction.value_tag)  # type: ignore
+        )
+
+    @staticmethod
+    def ignore_prediction(
+        product: Product,
+        category: str,
+    ):
+        # check whether this is new information or if the predicted category
+        # is not a parent of a current/already predicted category
+        return (
+            category in product.categories_tags
+            or CategoryImporter.is_parent_category(
+                category, set(product.categories_tags)
+            )
+        )
+
 
 class ProductWeightImporter(InsightImporter):
     @staticmethod
     def get_type() -> InsightType:
         return InsightType.product_weight
 
     @staticmethod
-    def group_by_subtype(insights: List[Insight]) -> Dict[str, List[Insight]]:
-        insights_by_subtype: Dict[str, List[Insight]] = {}
+    def get_required_prediction_types() -> Set[PredictionType]:
+        return {PredictionType.product_weight}
 
-        for insight in insights:
-            matcher_type = insight.data["matcher_type"]
-            insights_by_subtype.setdefault(matcher_type, [])
-            insights_by_subtype[matcher_type].append(insight)
+    @classmethod
+    def is_conflicting_insight(
+        cls, candidate: ProductInsight, reference: ProductInsight
+    ) -> bool:
+        return candidate.value == reference.value
 
-        return insights_by_subtype
+    @staticmethod
+    def group_by_subtype(predictions: List[Prediction]) -> Dict[str, List[Prediction]]:
+        predictions_by_subtype: Dict[str, List[Prediction]] = {}
 
-    def _generate_insights(
-        self,
-        product: Optional[Product],
-        barcode: str,
-        insights: List[Insight],
-        server_domain: str,
-    ) -> Iterator[Insight]:
-        if (
-            self.get_seen_count(barcode=barcode, server_domain=server_domain)
-            or (product and product.quantity is not None)
-            or not insights
-        ):
-            return
+        for prediction in predictions:
+            matcher_type = prediction.data["matcher_type"]
+            predictions_by_subtype.setdefault(matcher_type, [])
+            predictions_by_subtype[matcher_type].append(prediction)
 
-        insights_by_subtype = self.group_by_subtype(insights)
+        return predictions_by_subtype
 
-        insight_subtype = insights[0].data["matcher_type"]
+    @classmethod
+    def generate_candidates(
+        cls,
+        product: Product,
+        predictions: List[Prediction],
+    ) -> Iterator[ProductInsight]:
+        if (product and product.quantity is not None) or not predictions:
+            return
 
-        multiple_weights = False
+        insights_by_subtype = cls.group_by_subtype(predictions)
+        insight_subtype = predictions[0].data["matcher_type"]
```

Comment:
All insights are sorted by ascending `priority`, (see `generate_insights` method), and insights with subtype `with_mention` and `with_ending_mention` come first (priority = 1).
We should group insights by priority instead of by subtype anyway, as the current implementation has a bug (`insight_subtype != "with_mention"` -> `insight_subtype not in ("with_mention", "with_ending_mention")`.
I'm taking care of it